src/index.js

```diff
@@ -183,20 +183,56 @@ export default function (babel) {
 
 
   // HELPER FUNCTIONS
+  function visitTailExpressions(path, allowLoops, visitor) {
+    path.resync();
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+    let tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (let tailPath of tailPaths) {
+      if (tailPath.isExpressionStatement()) {
+        visitor(tailPath);
+      } else if (tailPath.isFunctionDeclaration()) {
+        // Promote function declarations to exprs.
```

Comment:
For the expression case, we can do the same thing that we're doing in `addImplicitReturns`, which is to put the name of the function in the arr.push / obj[k] =

so for example: 
```
{ for elem x in arr: x, f() -> x }
```
becomes
```
const _obj = {};
for (...) {
  const x = arr[_i];
  function f() { return x }
  _obj[x] = f;
}
```
thoughts?